[\# library(SVDFunctions)]{.c8 .c9 .c10}

[\# library(dplyr)]{.c8 .c9 .c10}

[]{.c8 .c9 .c10}

[\# The code below has been performed for GBR sampleset]{.c8 .c9 .c10}

### [\# Generate genotype matrix]{.c8} {#h.tro66a1ovkhb .c11}

[variants \<- read.table(\"one_consequence.tsv\", header=TRUE,
sep=\"\\t\") ]{.c1}

[1 476 344 rows #annotation file from vep]{.c1}

[vars \<- paste(paste(paste(\"chr\", variants\$chr, sep = \"\"),
variants\$pos, sep = \":\"), variants\$ref, variants\$alt, sep =
\"\\t\")]{.c1}

[samples \<- ]{.c7 .c10}[scan(\"intersected_samples.txt\", what =
character())]{.c3}

[read 91 samples \#]{.c7 .c10}[read samples]{.c7}

[result \<- SVDFunctions::scanBinaryFile("new_bin_bin", "new_bin_meta",
samples = samples, vars = vars, DP = 0, GQ = 0)]{.c1}

### [\# Generate merged file of annotaion and genotype distribution of samples]{.c8} {#h.1spxhibh1q8u .c11}

[variants \<- as.data.frame(cbind(vars, variants))]{.c1}

[master_table \<- inner_join(result, variants)]{.c1}

[161558 rows]{.c1}

[]{.c1}

### [\# ]{.c8}[Calculate the internal allele frequency]{.c8 .c9} {#h.dr6i13isbm6t .c11}

[]{.c1}

[master_table\$internal_af \<- (master_table\$het +
master_table\$hom_alt\*2) / 182 (n_samples\*2)]{.c1}

[]{.c1}

### [\# Calculate the variance between gnomAD and internal frequencies]{.c8 .c9} {#h.1wr04uuetbe2 .c11}

[]{.c1}

[master_table\$var \<- master_table\$internal_af -
master_table\$freq]{.c1}

### [\# Allele frequency filtering]{.c8 .c9} {#h.auub24wxwwbe .c11}

#### [\# ]{.c8}[Firstly, we can apply basic filters, dividing our data by variant annotation]{.c8 .c9} {#h.j955bh5q0z1c .c5}

[]{.c12 .c9}

[\# synonymous variants:]{.c8 .c9 .c10}

[master_syns \<- master_table %\>% filter(reason ==
"synonymous_variant")]{.c1}

[]{.c1}

[]{.c1}

[\# missense+ptv variants:]{.c15 .c10}

[master_mis_ptv \<- master_table %\>% filter(reason ==
\"missense_variant\" \| reason == \"frameshift_variant\" \| reason ==
\"stop_gained\" \| reason == \"stop_lost\" \| reason ==
\"splice_acceptor_variant\" \| reason == \"splice_donor_variant\" \|
reason == \"start_lost\")]{.c1}

[]{.c1}

#### [\# Then, we need to perform filtering and get variants we actually want to work with. It will include:]{.c8 .c9} {#h.np1tgjbq7hed .c5}

[]{.c9 .c12}

1.  [calculator variance between gnomAD allele frequency and internal
    allele frequency]{.c3}
2.  [filtering by \|variance\| \<= 2\*std of this variance]{.c3}
3.  [filtering by gnomAD allele frequency]{.c3}
4.  [filtering by internal allele frequency]{.c3}

[]{.c3}

[\# This should help in filtering discordant variants]{.c8 .c9}

[]{.c3}

[\# synonymous variants:]{.c8 .c9 .c10}

[]{.c1}

[master_syns_1p_filt \<- master_syns %\>% filter((freq \<= 0.01 & freq
!=0) & internal_af \<= 0.01 & (var \<= 2\*sd(master_table\$var) & var
\>= -2\*sd(master_table\$var))]{.c1}

[]{.c1}

[\# missense+ptv variants:]{.c8 .c9 .c10}

[]{.c1}

[master_mis_ptv_1p_filt \<- master_mis_ptv %\>% filter((freq \<= 0.01 &
freq !=0) & internal_af \<= 0.01 & (var \<= 2\*sd(master_table\$var) &
var \>= -2\*sd(master_table\$var))]{.c1}

### [\# Platform parameters]{.c8}[: DP=0, GQ=0, gnomAD_MAF = 0.01, отдельно syn, отдельно mis+ptv]{.c8} {#h.f7zhmq5b9x3 .c11}

### [\# Group variants by gene]{.c8} {#h.y96jo5l7kib8 .c11}

[\# synonymous variants:]{.c8 .c9 .c10}

[master_1p_syns_gr \<- master_syns_1p_filt %\>%                        
                                                                       
  group_by(gene) %\>%]{.c1}

[  summarise(]{.c1}

[    hom_ref = sum(hom_ref, na.rm = TRUE),]{.c1}

[    het = sum(het, na.rm = TRUE),]{.c1}

[    hom_alt = sum(hom_alt, na.rm = TRUE),]{.c1}

[    n_variants = sum(n_variants, na.rm = TRUE),]{.c1}

[    call_rate = mean(call_rate, na.rm = TRUE)]{.c1}

[  )]{.c1}

[]{.c1}

[\# missense+ptv variants:]{.c8 .c9 .c10}

[master_1p_mis_ptv_gr \<- master_mis_ptv_1p_filt %\>%                  
                                                                       
        group_by(gene) %\>%]{.c1}

[  summarise(]{.c1}

[    hom_ref = sum(hom_ref, na.rm = TRUE),]{.c1}

[    het = sum(het, na.rm = TRUE),]{.c1}

[    hom_alt = sum(hom_alt, na.rm = TRUE),]{.c1}

[    n_variants = sum(n_variants, na.rm = TRUE),]{.c1}

[    call_rate = mean(call_rate, na.rm = TRUE)]{.c1}

[  )]{.c1}

### [\# Import files from platform]{.c8 .c9} {#h.yvelsy3knchl .c11}

[\# synonymous variants:]{.c8 .c9 .c10}

[syns_controls_1p \<- read.table(\"syns_genes_controls_1p.tsv\",
header=TRUE, sep=\"\\t\")]{.c1}

[]{.c1}

[\# missense+ptv variants:]{.c8 .c9 .c10}

[mis_ptv_controls_1p \<- read.table(\"mis_ptv_genes_controls_1p.tsv\",
header=TRUE, sep=\"\\t\")]{.c1}

### [\# Merge cases matrix and controls matrix]{.c8 .c9} {#h.vzuqvtd623kv .c11}

[]{.c12 .c9}

[\# Cases samples get the flag '\_test' ]{.c8 .c9 .c10}

[]{.c8 .c9 .c10}

[\# synonymous variants:]{.c8 .c9 .c10}

[merged_data_syns_1p \<- merge(master_1p_syns_gr, syns_controls_1p, all
= TRUE, suffix = c(\'\_test\', \'\_control\'), by = \'gene\')]{.c1}

[]{.c1}

[merged_data_syns_1p\$hom_ref_control\[is.na(merged_data_syns_1p\$hom_ref_control)\]
\<-
merged_data_syns_1p\$n_variants_test\[is.na(merged_data_syns_1p\$hom_ref_control)\]
\* 496]{.c7 .c10}

[]{.c8 .c9 .c10}

[fisher.test(matrix(hom_ref\*2 + het, het + 2\* hom_alt,)]{.c1}

[]{.c1}

[\# missense+ptv variants:]{.c8 .c9 .c10}

[merged_data_mis_ptv_1p \<- merge(master_1p_mis_ptv_gr,
mis_ptv_controls_1p, all = TRUE, suffix = c(\'\_test\', \'\_control\'),
by = \'gene\')]{.c1}

[]{.c1}

[merged_data_mis_ptv_1p\[is.na(merged_data_mis_ptv_1p)\] = 0]{.c1}

### [\# Allele based fisher test]{.c8} {#h.t7cg86xnfnbw .c11}

[\# synonymous variants:]{.c8 .c9 .c10}

[merged_data_syns_fisher \<- merged_data_syns_1p %\>%]{.c3}

[  rowwise() %\>%]{.c3}

[  mutate(]{.c3}

[    fisher_p\_value = fisher.test(matrix(c(]{.c3}

[      (het_control + hom_alt_control \* 2),]{.c3}

[      (hom_ref_control \* 2 + het_control),]{.c3}

[      (het_test + hom_alt_test \* 2),]{.c3}

[      (hom_ref_test \* 2 + het_test)]{.c3}

[    ), nrow = 2, byrow = TRUE))\$p.value,]{.c3}

[    ]{.c3}

[    odds_ratio = fisher.test(matrix(c(]{.c3}

[      (het_control + hom_alt_control \* 2),]{.c3}

[      (hom_ref_control \* 2 + het_control),]{.c3}

[      (het_test + hom_alt_test \* 2),]{.c3}

[      (hom_ref_test \* 2 + het_test)]{.c3}

[    ), nrow = 2, byrow = TRUE))\$estimate)]{.c3}

[]{.c1}

[\# missense+ptv variants:]{.c8 .c9 .c10}

[merged_data_mis_ptv_fisher \<- merged_data_mis_ptv_1p %\>%]{.c3}

[  rowwise() %\>%]{.c3}

[  mutate(]{.c3}

[    fisher_p\_value = fisher.test(matrix(c(]{.c3}

[      (het_control + hom_alt_control \* 2),]{.c3}

[      (hom_ref_control \* 2 + het_control),]{.c3}

[      (het_test + hom_alt_test \* 2),]{.c3}

[      (hom_ref_test \* 2 + het_test)]{.c3}

[    ), nrow = 2, byrow = TRUE))\$p.value,]{.c3}

[    ]{.c3}

[    odds_ratio = fisher.test(matrix(c(]{.c3}

[      (het_control + hom_alt_control \* 2),]{.c3}

[      (hom_ref_control \* 2 + het_control),]{.c3}

[      (het_test + hom_alt_test \* 2),]{.c3}

[      (hom_ref_test \* 2 + het_test)]{.c3}

[    ), nrow = 2, byrow = TRUE))\$estimate]{.c3}

[  )]{.c7}

### [\# Calculate p value distributions for visualization]{.c8} {#h.l6mrwcfvzymr .c11}

[\# synonymous variants:]{.c8 .c9 .c10}

[merged_data_syns_fisher\$fisher_p\_value\[merged_data_syns_fisher\$fisher_p\_value
== 0\] \<- .Machine\$double.xmin]{.c1}

[]{.c1}

[merged_data_syns_fisher\$fisher_p\_value \<-
as.numeric(merged_data_syns_fisher\$fisher_p\_value)]{.c1}

[]{.c1}

[merged_data_syns_fisher\$neg_log_p \<-
-log10(merged_data_syns_fisher\$fisher_p\_value)]{.c1}

[]{.c1}

[merged_data_syns_fisher\$theoretical \<-
-log10((rank(merged_data_syns_fisher\$fisher_p\_value) - 0.5) /
length(merged_data_syns_fisher\$fisher_p\_value))]{.c1}

[]{.c1}

[\# missense+ptv variants:]{.c8 .c9 .c10}

[merged_data_mis_ptv_fisher\$fisher_p\_value\[merged_data_mis_ptv_fisher\$fisher_p\_value
== 0\] \<- .Machine\$double.xmin]{.c1}

[]{.c1}

[merged_data_mis_ptv_fisher\$fisher_p\_value \<-
as.numeric(merged_data_mis_ptv_fisher\$fisher_p\_value)]{.c1}

[]{.c1}

[merged_data_mis_ptv_fisher\$neg_log_p \<-
-log10(merged_data_mis_ptv_fisher\$fisher_p\_value)]{.c1}

[merged_data_mis_ptv_fisher\$theoretical \<-
-log10((rank(merged_data_mis_ptv_fisher\$fisher_p\_value) - 0.5) /
length(merged_data_mis_ptv_fisher\$fisher_p\_value))]{.c1}

[]{.c1}

### [\# Save the result table]{.c8} {#h.snnpvof8y8iz .c11}

[\# synonymous variants:]{.c15 .c10}

[write.table(merged_data_syns_fisher,
\"for_qpplot_carriers_syns_1p_new_p\_value.tsv\", sep=\"\\t\")]{.c1}

[]{.c1}

[\# missense+ptv variants:]{.c15 .c10}

[write.table(merged_data_mis_ptv_fisher,
\"for_qpplot_carriers_mis_ptv_1p_new_p\_value.tsv\", sep=\"\\t\")]{.c1}

[]{.c1}

### [\# ]{.c8}[Genotype based Fisher test]{.c8 .c9} {#h.hqsixfqnwc65 .c11}

[]{.c8 .c9 .c10}

[\# synonymous variants:]{.c15 .c10}

[fisher_p\_values_syns \<- numeric(nrow(]{.c7}[merged_data_syns_1p]{.c7
.c10}[))]{.c3}

[]{.c3}

[odds_ratios_syns \<- numeric(nrow(]{.c7}[merged_data_syns_1p]{.c7
.c10}[))]{.c3}

[]{.c3}

[total_control \<- 496]{.c3}

[]{.c3}

[total_gbr \<- 91]{.c3}

[]{.c3}

[for (i in 1:nrow(merged_data_syns_1p)) {]{.c3}

[  het_control \<- merged_data_syns_1p\$het_control\[i\]]{.c3}

[  het_test \<- merged_data_syns_1p\$het_test\[i\]]{.c3}

[  ]{.c3}

[  remaining_control \<- total_control - het_control]{.c3}

[  remaining_gbr \<- total_gbr - het_test]{.c3}

[    if (any(c(het_control, het_test, remaining_control, remaining_gbr)
\< 0)) {]{.c3}

[    next]{.c3}

[  }]{.c3}

[    matrix_data \<- matrix(c(het_control, remaining_control, het_test,
remaining_gbr), nrow = 2)]{.c3}

[    if (any(rowSums(matrix_data) == 0) \|\| any(colSums(matrix_data) ==
0)) {]{.c3}

[    next]{.c3}

[  }]{.c3}

[  fisher_test \<- fisher.test(matrix_data)]{.c3}

[  fisher_p\_values_syns\[i\] \<- fisher_test\$p.value]{.c3}

[  odds_ratios_syns\[i\] \<- ifelse(is.finite(fisher_test\$estimate),
fisher_test\$estimate, NA)]{.c3}

[}]{.c7}

[]{.c3}

[results_syns_genes_fisher \<- data.frame(]{.c3}

[  gene = ]{.c7}[merged_data_syns_1p]{.c7 .c10}[\$gene,  fisher_p\_value
= fisher_p\_values_syns,]{.c3}

[  odds_ratio = odds_ratios_syns)]{.c3}

[]{.c3}

[\# missense+ptv variants:]{.c15 .c10}

[fisher_p\_values_mis_ptv \<-
numeric(nrow(]{.c7}[merged_data_mis_ptv_1p]{.c7 .c10}[))]{.c3}

[]{.c3}

[odds_ratios_mis_ptv \<- numeric(nrow(]{.c7}[merged_data_mis_ptv_1p]{.c7
.c10}[))]{.c3}

[]{.c3}

[total_control \<- 496]{.c3}

[]{.c3}

[total_gbr \<- 91]{.c3}

[]{.c3}

[for (i in 1:nrow(merged_data_syns_1p)) {]{.c3}

[  het_control \<- merged_data_syns_1p\$het_control\[i\]]{.c3}

[  het_test \<- merged_data_syns_1p\$het_test\[i\]]{.c3}

[  ]{.c3}

[  remaining_control \<- total_control - het_control]{.c3}

[  remaining_gbr \<- total_gbr - het_test]{.c3}

[    if (any(c(het_control, het_test, remaining_control, remaining_gbr)
\< 0)) {]{.c3}

[    next]{.c3}

[  }]{.c3}

[    matrix_data \<- matrix(c(het_control, remaining_control, het_test,
remaining_gbr), nrow = 2)]{.c3}

[    if (any(rowSums(matrix_data) == 0) \|\| any(colSums(matrix_data) ==
0)) {]{.c3}

[    next]{.c3}

[  }]{.c3}

[  fisher_test \<- fisher.test(matrix_data)]{.c3}

[  fisher_p\_values_syns\[i\] \<- fisher_test\$p.value]{.c3}

[  odds_ratios_syns\[i\] \<- ifelse(is.finite(fisher_test\$estimate),
fisher_test\$estimate, NA)]{.c3}

[}]{.c3}

[]{.c3}

[results_mis_ptv_genes_fisher \<- data.frame(]{.c3}

[  gene = ]{.c7}[merged_data_mis_ptv_1p]{.c7 .c10}[\$gene,
 fisher_p\_value = fisher_p\_values_mis_ptv,]{.c3}

[  odds_ratio = odds_ratios_mis_ptv)]{.c3}

### [\# Calculate p value distributions for visualization]{.c8 .c9} {#h.1ll7nvroxw1u .c11}

[\# synonymous variants:]{.c15 .c10}

[results_syns_genes_fisher]{.c7}[\$fisher_p\_value\[]{.c7
.c10}[results_syns_genes_fisher]{.c7}[\$fisher_p\_value == 0\] \<-
.Machine\$double.xmin]{.c1}

[]{.c1}

[results_syns_genes_fisher]{.c7}[\$fisher_p\_value \<- as.numeric(]{.c7
.c10}[results_syns_genes_fisher]{.c7}[\$fisher_p\_value)]{.c1}

[]{.c1}

[results_syns_genes_fisher]{.c7}[\$neg_log_p \<- -log10(]{.c7
.c10}[results_syns_genes_fisher]{.c7}[\$fisher_p\_value)]{.c1}

[]{.c1}

[results_syns_genes_fisher]{.c7}[\$theoretical \<- -log10((rank(]{.c7
.c10}[results_syns_genes_fisher]{.c7}[\$fisher_p\_value) - 0.5) /
length(]{.c7
.c10}[results_syns_genes_fisher]{.c7}[\$fisher_p\_value))]{.c1}

[]{.c1}

[\# missense+ptv variants:]{.c15 .c10}

[results_mis_ptv_genes_fisher]{.c7}[\$fisher_p\_value\[]{.c7
.c10}[results_mis_ptv_genes_fisher]{.c7}[\$fisher_p\_value == 0\] \<-
.Machine\$double.xmin]{.c1}

[]{.c1}

[results_mis_ptv_genes_fisher]{.c7}[\$fisher_p\_value \<-
as.numeric(]{.c7
.c10}[results_mis_ptv_genes_fisher]{.c7}[\$fisher_p\_value)]{.c1}

[]{.c1}

[results_mis_ptv_genes_fisher]{.c7}[\$neg_log_p \<- -log10(]{.c7
.c10}[results_mis_ptv_genes_fisher]{.c7}[\$fisher_p\_value)]{.c1}

[]{.c1}

[results_mis_ptv_genes_fisher]{.c7}[\$theoretical \<- -log10((rank(]{.c7
.c10}[results_mis_ptv_genes_fisher]{.c7}[\$fisher_p\_value) - 0.5) /
length(]{.c7
.c10}[results_mis_ptv_genes_fisher]{.c7}[\$fisher_p\_value))]{.c1}

[]{.c1}

### [\# Save the result table]{.c8} {#h.3zdjoe3vtc64 .c11}

[\# synonymous variants:]{.c10 .c15}

[write.table(]{.c7 .c10}[results_syns_genes_fisher]{.c7}[,
\"for_qpplot_genes_syns_1p_new_p\_value.tsv\", sep=\"\\t\")]{.c1}

[]{.c1}

[\# missense+ptv variants:]{.c15 .c10}

[write.table(]{.c7 .c10}[results_mis_ptv_genes_fisher]{.c7}[,
\"for_qpplot_genes_mis_ptv_1p_new_p\_value.tsv\", sep=\"\\t\")]{.c1}

### [QQPlots]{.c8} {#h.o5x2uyihyspj .c11}

[]{.c1}

[import pandas as pd ]{.c1}

[import numpy as np ]{.c1}

[import scipy.stats as stats ]{.c1}

[import matplotlib.pyplot as plt ]{.c1}

[]{.c1}

[results_df = pd.read_csv(\"]{.c7 .c10}[input table]{.c15 .c10}[\" ,sep
= \"\\t\") ]{.c1}

[print(results_df) ]{.c1}

[expected = results_df\[\'theoretical\'\] ]{.c1}

[observed = results_df\[\'neg_log_p\'\]]{.c1}

[plt.figure(figsize=(6, 6)) ]{.c1}

[plt.scatter(expected, observed, edgecolor=\'black\')
plt.plot(\[expected.min(), expected.max()\], \[expected.min(),
expected.max()\], color=\'red\', linestyle=\'\--\') ]{.c1}

[plt.xlabel(\"Expected -log10(p-value)\") ]{.c1}

[plt.ylabel(\"Observed -log10(p-value)\") ]{.c1}

[plt.title(\"QQ-plot of p-values\") ]{.c1}

[plt.show()]{.c7 .c10}
